Add features for the backend in the full-stack capstone project: Tesla Car Test Drive Booking

ivangonzalez224 opened the pull request:
## Final PR - API app for Full Stack Capstone project - Tesla Car Test Drive Booking

Link to the front end app Pull Request: https://github.com/ivangonzalez224/Tesla-car-booking-front-end/pull/35

**Description:**

This pull request implements the following features for the Tesla car test drive booking API app:

**Database and schema:**
- Created the users, cars, and reservations tables.
- Added the seed data for cars and users.

**Routing:**
- Created the routes to sign in and sign out.
- Created the routes to create and delete a car.
- Created the routes to list and create a reservation.

**Authentication:**
- Added the the active_model_serializers and devise-jwt gems for user authentication from frontend to backend.
- Created the SessionsController for user authentication.
- Added User model with Devise authentication and JWT.
- Only admin users can access "Add car" and "Delete car" functionalities.

**Models and Controllers:**
- Created the user, car and reservation models including validations.
- Create the user, car and reservation API controllers including actions.

**API Documentation:**
- Included rswag-ui and rswag-api gems.
- Added Rswag API and Rswag UI configuration.

**Technical Setup:**

Gitflow workflow implemented.
Postgres database used.

Meltrust commented on spec/models/user_spec.rb:
- [ ] In order for this spec file to work, you need to include the   `gem 'shoulda-matchers', '~> 3.1'` in your gemfile.  Kindly include it so your tests work.  Otherwise, rspec throws an error. 👍 

![image](https://github.com/hafiz1379/Tesla-car-booking-back-end/assets/57421823/fd0b15fa-024a-44d3-a1c9-7489f869a5a3)

Meltrust commented on spec/models/reservation_spec.rb:
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe Reservation, type: :model do
+  pending "add some examples to (or delete) #{__FILE__}"
+end
```
- [ ] This file has no implemented examples.  Kindly remove it if not using it. 👍

Meltrust commented on README.md:
```diff
@@ -1 +1,184 @@
+
+In this project that used RoR to build a Tesla Car booking api.
+
+### Prerequisites
```
- [ ] Kindly, include in your readme file a "tests" section with specific instructions about how to run your tests. 👍

Meltrust commented on README.md:
```diff
@@ -1 +1,184 @@
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+## 📑 Database ERD <a name="erd"></a>
```
- [ ]  You are using Rspec, but the "test" folder still exists, which is used by another test framework. Kindly consider deleting that folder if you are not using it. 👍

Meltrust commented on README.md:
```diff
@@ -1 +1,184 @@
+
+## 📑 Database ERD <a name="erd"></a>
+
+<img src='app/assets/images/table-tesla-car-booking.png' />
```
- [ ] Kindly, include in your instructions the required db commands that your users will need to run your backend app: ``rails db:create, rails db:migrate, rails db:seed``. 👍

Meltrust commented on app/controllers/api/v1/cars_controller.rb:
```diff
@@ -0,0 +1,44 @@
+    end
+  end
+
+  def destroy
+    @car = Car.find(params[:id])
+    @car.update(is_removed: true)
+  end
```
- [ ] What happens here if the car is not successfully destroyed?  Kindly, add some error handling just like you did in the update method (line 25). 👍

Meltrust commented on app/controllers/api/v1/reservations_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+  end
+
+  def destroy
+    @reservation.destroy
+    head :no_content
+  end
```
- [ ] What happens here if the reservation is not successfully destroyed?  Kindly, add some error handling just like you did in the update method (line 29). 👍

Meltrust reviewed (changes requested):
Hi,

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation, but you are almost there!









#### You are really close to finishing the Microverse program!! Keep it up! 👍👍👍

![You can do it](https://media.giphy.com/media/FemTGpEZeK27M5fKL4/giphy.gif)

### After implementing the requested changes, please submit another review request. ♻️
_Check the comments under the review._







Cheers and Happy coding!👏👏👏

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the previous reviews unless it is requested otherwise._**

------

ivangonzalez224 commented:
Hi @Meltrust thank you for the review:
Changes done: 
- updated the destroy method in the cars and reservations controllers.
- included the gem shoulda-matchers.
- removed the test folder from the project.
- updated the readme file by adding the instructions to run the database and the tests.

Meltrust reviewed (approved):
# **Wow, you did it 🎉**
![Brilliant](https://media.giphy.com/media/d31w24psGYeekCZy/giphy.gif)
## Thank you for the changes implemented 💪 🥇 ㊗️ 
#### _Unless you want to add more features, go ahead to your final presentation ⏩ ⏩ ⏩_ 
## You are about to finish the Microverse program.  You have come a long way!!!




### Good luck in the software industry!!  I'll see you there. ✨
### **Congratulations!!!!!!**  🎉
![applause](https://media.giphy.com/media/srg19CG0cKMuI/giphy.gif)

### To highlight


- Superb job ✔️  



Cheers and Happy coding!👏👏👏


------

hafiz1379 commented:
@Meltrust 
Thank you for investing time in reviewing our project and providing valuable feedback!
